File: src/lib/dev-auth.ts
import { prisma } from "@/lib/db";
import { createSessionToken, setSessionCookie } from "@/lib/session";

const IS_DEV_BYPASS =
  process.env.ENVIRONMENT === "development" &&
  (!process.env.WORKOS_API_KEY || process.env.WORKOS_API_KEY === "sk_live_...");

/**
 * Returns true if dev auth bypass is active
 * (ENVIRONMENT=development and no real WorkOS key).
 */
export function isDevBypass(): boolean {
  return IS_DEV_BYPASS;
}

/**
 * Get or create the dev seed user and set the session cookie.
 * Call this from the SSO route when dev bypass is active.
 */
export async function devLogin() {
  // Ensure org exists
  let org = await prisma.organization.findFirst();
  if (!org) {
    org = await prisma.organization.create({
      data: {
        name: "Dev Organization",
        slug: "dev-org",
        billingEmail: "dev@efilo.ai",
      },
    });
  }

  // Ensure dev user exists
  let user = await prisma.user.findUnique({
    where: { email: "dev@efilo.ai" },
  });
  if (!user) {
    user = await prisma.user.create({
      data: {
        email: "dev@efilo.ai",
        name: "Dev User",
        role: "ADMIN",
        organizationId: org.id,
        lastLoginAt: new Date(),
      },
    });
  } else {
    user = await prisma.user.update({
      where: { id: user.id },
      data: { lastLoginAt: new Date() },
    });
  }

  const token = await createSessionToken({
    userId: user.id,
    email: user.email,
    role: user.role,
  });

  await setSessionCookie(token);

  return user;
}

File: backend/app/auth/dev_auth.py
"""Development bypass authentication.

When ENVIRONMENT=development and no real WorkOS key is configured,
auto-login as the seed dev user without any IdP interaction.
"""

from datetime import datetime, timezone

from sqlalchemy import select

from app.config import get_settings
from app.db.session import async_session_factory
from app.models.enums import AuthMethod, UserRole
from app.models.organization import Organization
from app.models.user import User

settings = get_settings()

IS_DEV_BYPASS = settings.is_development and (
    not settings.workos_api_key
    or settings.workos_api_key == "sk_live_..."
    or settings.workos_api_key.startswith("sk_test_")
)


async def dev_login() -> User:
    """Create or fetch the dev user, update lastLoginAt, return User."""
    async with async_session_factory() as session:
        # Ensure at least one Organization exists
        result = await session.execute(select(Organization).limit(1))
        org = result.scalar_one_or_none()
        if not org:
            org = Organization(
                name="Dev Organization",
                slug="dev",
                billing_email="dev@efilo.ai",
            )
            session.add(org)
            await session.flush()

        # Ensure dev user exists
        result = await session.execute(
            select(User).where(User.email == "mnayyar@efilo.ai")
        )
        user = result.scalar_one_or_none()
        if not user:
            user = User(
                email="mnayyar@efilo.ai",
                name="Mateen Nayyar",
                role=UserRole.ADMIN,
                auth_method=AuthMethod.EMAIL_PASSWORD,
                organization_id=org.id,
            )
            session.add(user)
        else:
            user.last_login_at = datetime.utcnow()

        await session.commit()
        await session.refresh(user)
        return user
